runtime/include/sandbox_set_as_running_user.h

```diff
@@ -42,9 +42,11 @@ sandbox_set_as_running_user(struct sandbox *sandbox, sandbox_state_t last_state)
 	runtime_sandbox_total_increment(SANDBOX_RUNNING_USER);
 	runtime_sandbox_total_decrement(last_state);
 
-	/* WARNING: This state change needs to be at the end of this transition because all code below this assignment
-	 * is preemptable */
+	/* WARNING: All code below this assignment is preemptable */
 	sandbox->state = SANDBOX_RUNNING_USER;
+
+	/* Now that we are preemptable, we can replay deferred sigalrms */
+	software_interrupt_deferred_sigalrm_replay();
```

Comment:
You'll need a `volatile asm("")` above this line, I believe. This is a compiler barrier that disallows the compiler to reorder the setting of state after this function.

Reply:
I defined a Linux-like macro for this and added the barrier to both sides of the state transitions that enable / disable preemption. Might be overkill, but I assume code migration across the state assignment from either direction is dangerous.

Reply:
Both sides isn't necessary. When disabling preemptions, you want to make sure that accesses in the runtime aren't reordered before the disable. When enabling preemptions, you want to make sure that previous runtime code is not reordered after the enable. I'd rather encourage people not overusing concurrency primitives, here.

Reply:
Ok nevermind. Got my answer to the question above.